Implementation for Amazon DynamoDB

nicktacular opened the pull request:
Added an implementation for Amazon DynamoDB. Using the `$options` key to the constructor, you can set which `$storageName` uses which key name for access. For example, if your key-val store called `Awesome` uses a key called `MyKey` then you can set it like so:

```php
$credentials = new \Aws\Credentials\Credentials('yo', 'sup');

$aws = new \Aws\Sdk([
    'timeout' => 1,
    'credentials' => $credentials,
    'region' => 'us-west-2',
    'version' => 'latest',
    'endpoint' => 'http://127.0.0.1:11888',
]);

$storage = new AmazonDynamoDbStorage(
    $aws->createDynamoDb(),
    null,
    ['storage_keys' => ['Awesome' => 'MyKey']]
);
```

Otherwise, the option `default_key_name` which defaults to `Id` will be used as the key name.

nicktacular commented:
Oh, also fixes #41.

beberlei commented:
@nicktacular is the storage thing sometihng we could better configure in the mappings, i.e. annotations xml and such?

Otherwise this looks really great, what do you think @EmanueleMinotto ?

EmanueleMinotto commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,210 @@
+    private $options = [
+        'default_key_name' => 'Id',
+        'storage_keys' => [],
+    ];
```
Probably would be better to declare and define the properties `private $defaultKeyName = 'Id'` and `private $storageKeys = []`, for some reasons:
1. `$options` is used only by `AmazonDynamoDbStorage` so its content don't need to be flexible
2. if the final developer defines the *default_key_name* key as a non-scalar type, the class will generates different unexpected errors
3. a more strict type checking could be used asking for an `array $storageKeys = []`

What do you think about this @nicktacular?

nicktacular replied:
Agreed, will implement strict checking as well as separate these into different arrays.

EmanueleMinotto commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,210 @@
+     */
+    public function getName()
+    {
+        return 'amazondynamodb';
```
I would use `amazon_dynamodb` here, to reproduce a sort of *[vendor]*_*[library]* like in the other implementations.

nicktacular replied:
Agreed, will fix this.

EmanueleMinotto commented on tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php:
```diff
@@ -0,0 +1,188 @@
+{
+    private function getDynamoDbMock($methods = [])
+    {
+        $client = $this->getMockBuilder(DynamoDbClient::class)->disableOriginalConstructor();
```
Use `'Aws\DynamoDb\DynamoDbClient'` instead of `DynamoDbClient::class` please, so we can keep the php 5.4 support also on dev. :)

The same for every other `::class` used below.

nicktacular replied:
Ah yes, will replace with classname as a string instead.

EmanueleMinotto commented:
@nicktacular I'm not sure if I'm understanding the part:
> Using the $options key to the constructor, you can set which $storageName uses which key name for access.

In this moment, looking the implementation, it seems a lot like the [`DoctrineCacheStorage::flattenKey`](https://github.com/doctrine/KeyValueStore/blob/master/lib/Doctrine/KeyValueStore/Storage/DoctrineCacheStorage.php#L62-L74), can you provide a more concrete example please?
It could be useful for documentation too.

nicktacular commented:
@EmanueleMinotto Sorry, I wasn't very clear there. What I meant to say is that the default name of the key is configurable. Also, if you want to override that key for any particular table (i.e. `$storageName`) then you can do that as well. Once I fix the format of the constructor and the way this is stored, I think it'll be clearer.

nicktacular commented:
I believe I've addressed all the concerns except for this comment from @beberlei:

> is the storage thing sometihng we could better configure in the mappings, i.e. annotations xml and such?

I agree that perhaps this would be a good place handle this. Can you give me an example of what you'd like the annotation to look like?

Also, I'm wondering if I should write an integration test for an end-to-end test, configurable via PHPUnit with `<var name="DOCTRINE_KEYVALUE_AMAZON_DYNAMODB_ENDPOINT" value="http://localhost:8000" />` (among other properties) where, if you wish, I can write this test to work against some endpoint. One example could be to use [DynamoDB local](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tools.DynamoDBLocal.html) for this and skip those tests when this isn't configured. Thoughts on this?

beberlei commented:
@nicktacular Integration test sounds great that only runs when all required environment variables are set, use env variables for this please, because other tools can do that more easily.

EmanueleMinotto commented:
@nicktacular what's the status of the e2e test addition?

Ocramius commented on composer.json:
```diff
@@ -2,7 +2,8 @@
         "php": ">=5.5",
-        "doctrine/common": "^2.4"
+        "doctrine/common": "^2.4",
+        "aws/aws-sdk-php": "^3.8"
```
must be `require-dev`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+            $this->setDefaultKeyName($defaultKeyName);
+        }
+
+        if (!empty($tableKeys)) {
```
No need for this check

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @throws \InvalidArgumentException When the key name is invalid.
+     */
+    protected function validateKeyName($name)
```
`private`

nicktacular commented:
Sorry for the delay. Working on some issues with work; will be back to completing these tests shortly.

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    protected function validateKeyName($name)
+    {
+        if (!is_string($name)) {
+            throw new \InvalidArgumentException(
```
Please throw specialized exceptions, not SPL exceptions (write a new exception class - you may look at http://rosstuck.com/formatting-exception-messages/ )

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @throws \InvalidArgumentException When the name is invalid.
+     */
+    protected function validateTableName($name)
```
`private`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @throws \InvalidArgumentException When the key name is invalid.
+     */
+    public function setDefaultKeyName($name)
```
`private`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @return string The default key name.
+     */
+    public function getDefaultKeyName()
```
Needed?

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @throws \InvalidArgumentException When the key or table name is invalid.
+     */
+    public function setKeyForTable($table, $key)
```
`private`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @return string
+     */
+    public function getKeyNameForTable($tableName)
```
needed?

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @return array The key in DynamoDB format.
+     */
+    protected function prepareKey($storageName, $key)
```
`private`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @return bool
+     */
+    public function supportsPartialUpdates()
```
add an `{@inheritDoc}`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+     */
+    public function supportsPartialUpdates()
+    {
+        //This is not true, but partial updates are too complicated given the available interface,
```
This comment should be at method level

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Determine if the storage supports updating only a subset of properties,
```
Don't duplicate the parent docblock

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Does this storage support composite primary keys?
```
Don't duplicate the parent docblock, use `{@inheritDoc}`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Does this storage require composite primary keys?
```
Don't duplicate the parent docblock, use `{@inheritDoc}`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    public function insert($storageName, $key, array $data)
+    {
+        $this->client->putItem([
+            'TableName' => $storageName,
```
Consider using class constants for these - makes it easier to refactor and to understand/document the meaning

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Insert data into the storage key specified.
```
Don't duplicate the parent docblock, use `{@inheritDoc}`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Update data into the given key.
```
Don't duplicate the parent docblock, use `{@inheritDoc}`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Delete data at key.
```
Don't duplicate the parent docblock, use `{@inheritDoc}`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    public function update($storageName, $key, array $data)
+    {
+        //We are using PUT so we just replace the original item
+        $this->insert($storageName, $key, $data);
```
What if the key doesn't exist?

nicktacular replied:
If an item by that key does not exist, it is simply inserted. Is that your question?

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Find data at key.
```
Don't duplicate the parent docblock, use `{@inheritDoc}`

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+
+        $item = $item->get('Item');
+        if (!is_array($item)) {
+            throw new NotFoundException();
```
Maybe more customized exception message in this case?

nicktacular replied:
Actually, I don't even know if this is possible. I can't find whether or not this type of response is possible from AWS for DynamoDb, so I'm inclined to just remove this check. Thoughts?

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+        ]);
+
+        if (!$item) {
+            throw new NotFoundException();
```
Maybe include the key in the exception? Otherwise, hard to debug

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Return a name of the underlying storage.
```
Don't duplicate the parent docblock, use `{@inheritDoc}`

Ocramius commented on tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    /**
+     * @expectedException \InvalidArgumentException
```
Please use `$this->setExpectedException()` instead, so you know exactly at which point in the execution an exception should be thrown

Ocramius commented on tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php:
```diff
@@ -0,0 +1,278 @@
+
+use Doctrine\KeyValueStore\Storage\AmazonDynamoDbStorage;
+
+class AmazonDynamoDbTest extends \PHPUnit_Framework_TestCase
```
Add `@covers \Doctrine\KeyValueStore\Storage\AmazonDynamoDbStorage` to avoid coverage leakage

nicktacular commented:
Thanks for the feedback @Ocramius; I'll implement.

Ocramius commented on tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php:
```diff
@@ -0,0 +1,278 @@
+<?php
+
```
License headers

EmanueleMinotto replied:
@nicktacular after a rebase you can just run `php-cs-fixer fix`

Ocramius commented on tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    /**
+     * @expectedException \InvalidArgumentException
```
same as above

Ocramius commented on tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    /**
+     * @dataProvider invalidTableNames
```
same as above

Ocramius commented on tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php:
```diff
@@ -0,0 +1,278 @@
+
+    /**
+     * @dataProvider invalidTableNames
+     * @expectedException \InvalidArgumentException
```
same as above

Ocramius commented on tests/Doctrine/Tests/KeyValueStore/Storage/AmazonDynamoDbTest.php:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    /**
+     * @expectedException \Doctrine\KeyValueStore\NotFoundException
```
same as above

Ocramius commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,311 @@
+use Aws\DynamoDb\Marshaler;
+use Doctrine\KeyValueStore\NotFoundException;
+
+class AmazonDynamoDbStorage implements Storage
```
can be `final` if all the API is defined in `Storage`

nicktacular replied:
Not sure I quite follow. By definition, since this is not abstract, the API must be fully honored, but I don't think it's a good idea to make an implementation final if someone wants to extend that behavior. Perhaps I'm not understanding your request?

Ocramius commented:
@nicktacular thanks for the patience, and sorry for the noise :-) Just ordinary administration here

nicktacular commented:
@EmanueleMinotto I didn't realize there was a deadline. What's the time you need this done by?

EmanueleMinotto commented:
@nicktacular no deadline, I'm just trying to reach a RC before the end of the year :)
https://github.com/doctrine/KeyValueStore/milestones

nicktacular commented:
@EmanueleMinotto - the eventual consistency model is proving to be a significant hurdle in the functional tests. I will continue working on this, but wanted to let you know.

EmanueleMinotto commented:
@nicktacular don't worry, that's not a problem, thank you for your effort (ping me on IRC if you need a help) :)

EmanueleMinotto commented:
@nicktacular do you have news about this? If you can't, than I could merge this into a separate branch and complete it by myself :)

nicktacular commented:
@EmanueleMinotto I've been really sick recently. I also couldn't get integration tests to work, so I'm going to implement using DynamoDb local store instead. I will let you know as soon as I'm done.

EmanueleMinotto commented:
oh ok, no problem
thanks @nicktacular, let me know :)

nicktacular commented:
Rewrote functional tests to utilize DynamoDb local (removed waits and such), but now it seems to be randomly failing, possibly to do with something on OS X as per [this AWS thread](https://forums.aws.amazon.com/thread.jspa?threadID=157820). I'm going to bring up an Ubuntu machine and try there.

nicktacular commented:
Ok so it's not Dynamo. It's `BasicCrudTestCase` class which seems to be taking an inordinate amount on time on something. It's approximately 17 seconds per test and it seems to be hung up on the call:

```php
$this->manager = $this->createManager($this->storage, $mappingDriver);
```

For what reason I can't understand, but I'm going to dig into this now to see why this is happening. Updates to follow...

nicktacular commented:
Well, that was quick. When commenting out the `xml` mapping driver, the test speeds up immensely. So, I'm going to hazard a guess that the XSD reference in the `Post` fixture for XML mapping is what's causing this hangup:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://doctrine-project.org/schemas/orm/doctrine-mapping
                          http://raw.github.com/doctrine/doctrine2/master/doctrine-mapping.xsd">
	<entity name="Doctrine\Tests\KeyValueStore\Functional\Post" storage-name="post">
		<id>id</id>
	</entity>
</doctrine-mapping>
```

nicktacular commented:
Ok, so I've discovered a conceptual issue with tables in DynamoDb. How do we want to deal with mismatches on types for ids? Suppose, for example, the type is a string id in DynamoDb and the id is set to an integer. Should the type of the primary key be deduced by means of a `describeTable` operation on the table first, or provide it as a configuration option?

When this occurs, the raw exception from DynamoDb looks like this:

```
Client error: `POST http://localhost:8000` resulted in a `400 Bad Request` response:
{"__type":"com.amazon.coral.validate#ValidationException","message":"Type mismatch for attribute to update"}
```

Doing a `describeTable` operation seems wasteful, so I'd prefer to rely on configuration (i.e. annotation) instead. Any thoughts on this?

nicktacular commented:
Having thought about this more, I think that the right approach is to use an appropriate exception message to indicate that there is an attribute mismatch and leave it up to the consumer for resolving this problem. Additionally, a new configuration will need to be set to indicate the type of key that is on the table (or keys).

Furthermore, an additional test will need to be added to test composite keys which is the error @tiagobrito mentions. **However**, it's only feasible to deal with this error when the table and entity actually have composite keys. If there is a mismatch, it will still be up to the developer to address this issue and ensure that the table key schema and entity match.

Finally, I'm not convinced that these configuration options make sense to be annotations. These configurations are too specific to the adapter, so I think they should be adapter configurations and not moved to annotations.

In summary:

 * Will add configuration options so that the developer can configure the key or keys and the type for each key (string vs numeric) and the associated tests.
 * Will add functional tests for composite keys.
 * Will add functional tests for mismatches which will *necessarily* throw an exception with an adequate message explaining why this is an error and what is needed to address.

I will wait a few days before I continue this implementation in case there are any serious concerns with this approach.

EmanueleMinotto commented:
well done @nicktacular, I'm also against the usage of a new annotation for a single driver, so :+1: for:

> deduced by means of a describeTable operation on the table first

or other configurations you think are required for the DynamoDB storage

larowlan commented on lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php:
```diff
@@ -0,0 +1,294 @@
+    {
+        $this->client->putItem([
+            self::TABLE_NAME_KEY => $storageName,
+            self::TABLE_ITEM_KEY => $this->prepareKey($storageName, $key) + $this->marshaler->marshalItem($data),
```
DynamoDb chokes on empty values.
E.g 
```php
$data = [
  'foo' => '',
  'bar' => 'baz',
];
```
Would choke. The [existing implementation](https://github.com/doctrine/KeyValueStore/blob/master/lib/Doctrine/KeyValueStore/Storage/DynamoDbStorage.php#L194) includes a `prepareData` method - however it only consider one-dimensional arrays, multi-dimensional cleaning is needed too.

e.g
```php
$data = [
  'foo' => [
     'baz' => '',
     'bar' => 'boo',
  ],
  'bar' => 'baz',
];
```
needs to be sent as
```php
$data = [
  'foo' => [
     'bar' => 'boo',
  ],
  'bar' => 'baz',
];
```

larowlan commented:
Promised work is in https://github.com/nicktacular/KeyValueStore/pull/2, sent against @nicktacular branch so once combined will be available here

larowlan commented:
Let me know if you'd prefer to close this and open a new one instead of sending fixes against @nicktacular branch
